### ISSUE TITLE
Install to computer instead of user profile

### ISSUE THREAD

**Issue by rbcmhelpdesk** [atom/atom#5259]
We would like a package that installs to the computer instead of into the users profile.   Security policies prevent execution of programs installed to the profile and this will make the software available to all users of the computer.

**Comment by kevinsawicki** [comment]
The install location is based on the `UserProfile` environment variable so just change that variable to a location that is acceptable for your security policy.

**Comment by rbcmhelpdesk** [comment]
Hi Kevin,

 

When we launch the installation package we are not given a choice of installation locations.  

 

Group Policy for our organization blocks all executable files that are not installed in C:\Program Files 

 

Doug

**Comment by paulcbetts** [comment]
Hey Doug, 

Sorry, Atom will only install to a user's profile directory, you should change your GP settings so that users can execute code from their own folders. In future versions, a Program Files-compatible version might exist but that's a long way off.

**Comment by Zireael07** [comment]
How about just giving people an option where to install? It's pretty standard...

**Comment by rbcmhelpdesk** [comment]
When we run the installation program there is no point where there is a choice on where to install it.

 

If you know of a switch or something similar we could use that would be very helpful

 

Thanks,

**Comment by paulcbetts** [comment]
> How about just giving people an option where to install? It's pretty standard...

Nope, we won't do that, sorry.

**Comment by batjko** [comment]
In future versions, a Program Files-compatible version might exist but that's a long way off.

Why is that exactly? I was surprised to see it when the install location changed to the user directory. It's unusual and confusing to most Windows users.

What's the difficulty to just put it either into the Program Files, or any random location of the user's choice?

**Comment by 50Wliu** [comment]
I believe installing to Program Files requires administrator permissions, which is a no no.

>* No UAC dialogs, which means....
* ...installs to the local user account (i.e. under %LocalAppData%)

**Comment by paulcbetts** [comment]
> What's the difficulty to just put it either into the Program Files, or any random location of the user's choice?

If we put it in Program Files, we can no longer update silently, since changing those files requires us to pop a UAC dialog. This also prevents users who aren't Administrator from installing the app at all. 

I do want to solve the case that @rbcmhelpdesk describes, where they want to provision an entire lab with Atom and manage it via GP, but I want that to be a special-case scenario, not the mainstream one.

**Comment by thx1200** [comment]
Disappointing.  Yet another OSS that doesn't respect the Windows security model.  I can't wait until somebody blames Microsoft because a vulnerability is exploited by allowing users to patch/overwrite application EXE since it's installed to the user's data folder instead of the application folder.

**Comment by paulcbetts** [comment]
@thx1200 [UAC is not a security boundary](http://blogs.msdn.com/b/e7/archive/2009/02/05/update-on-uac.aspx)

**Comment by thx1200** [comment]
UAC is not a security boundary, but Program Files is, and Atom won't install to Program Files because, apparently, UAC.  Probably laziness, but okay, UAC.

**Comment by 50Wliu** [comment]
@thx1200 Google Chrome doesn't ask for UAC either.
I personally find UAC annoying and it's nice when a program doesn't ask for it when installing (or the few that require it *every single startup*).

**Comment by thx1200** [comment]
I can tell this is a losing battle because nobody here apparently understands enterprise security in a Windows domain.  That's fine, there are plenty of other text editors out there that don't install to %appdata%.  And, you are right, Google Chrome does also install to %AppData% OR %ProgramFiles% if you want it to, which Atom does not.  User choice.  Imagine that!  http://www.google.com/intl/en/chrome/business/browser/

**Comment by kevinsawicki** [comment]
@thx1200 There is a .zip Windows distribution which you can install to the location that makes sense for you. It is available as `atom-windows.zip` on https://github.com/atom/atom/releases/latest

**Comment by 50Wliu** [comment]
Yes, but remember that Squirrel is still in its infancy **and** that @paulcbetts has indicated that a Program Files-compatible version will be available eventually.

**Comment by batjko** [comment]
Also, let's try and stay constructive and friendly. No need to wrap valid
concerns into passive-aggressive woohaa.

**Comment by paulcbetts** [comment]
So, @thx1200 is technically right in that installing to Program Files is a bit more secure than installing to the User's directory - though I'd argue that if someone is in a position to overwrite your user files with arbitrary data, you are already in a very bad position, one that isn't worth defending; they are already on the other side of the airtight hatch. 

We've intentionally made this decision to be slightly less secure, in exchange for a number of benefits:

* Users that don't have Admin privileges at all can install Atom (i.e. in a Corp domain)
* We do not have to show UAC dialogs to update the app, so that we can do background updates
* We do not have to show UAC dialogs to install the app either
* We do not have to require the user to reboot in order to install an app update, because our files in Program Files are in-use by the currently running app.

**Comment by shawnr13** [comment]
May I make a suggestion?  With all the recent issues of cryptolocker, et al, I'm testing and hoping to deploy soon a GPO that will block many locations from allowing "arbitrary" .exe files to execute.  This includes the %localappdata%\*.exe and one path deeper, %localappdata%\*\*.exe.  So far the only real program blocked by that is Spotify, and users have been up in arms over it for years without a resolution.  I have seen many prorgrams (Chrome included) that run two folders deep.  Inside of AppData local it goes AppData\Local\CompanyName\ProductName\product.exe.  This is allowed on our network.  I have yet to see any viruses hide 2 folders deep.  Could you just add an extra layer (say, GitHub\Atom) to the install path?  I got the idea for the software restriction policies from many suggestions of the same paths all over the internet, so I know that many other organizations have similar SRPs.  Just a thought.  It might not help @rbcmhelpdesk that much, but it would at least help others with slightly less restrictive policies.

**Comment by paulcbetts** [comment]
This is 100% fixed in the latest version of Squirrel, see these two PRs for more details:

* https://github.com/Squirrel/Squirrel.Windows/pull/337
* https://github.com/Squirrel/Squirrel.Windows/pull/308

**Comment by shawnr13** [comment]
Thanks @paulcbetts.  I used the --machine argument and it worked flawlessly.  Just one last suggestion... notating that somewhere on the download page (I understand if you just want atomsetup.exe to install with zero interaction and not adding it as an option).

**Comment by mattwolfgang** [comment]
"you should change your GP settings so that users can execute code from their own folders"

Ha, you're funny! Feel free to run as you please, CryptoLocker!

**Comment by stevesobol** [comment]
May I suggest that silent updates suck?

As a user, I want to know when a program is updating itself. You want to talk security? Hiding installs/updates/changes is something that malware does.

I'd also suggest that telling people to change group policy in a way that makes their domains and computers less secure is... ok, I'll be blunt... obnoxious. And wrong. Especially when the decisions you've already made are allegedly related to "security."

It wouldn't be the first time an OSS author or group of authors has made lousy security decisions and when the community complained, told them to go pound salt. (Filezilla FTP client and plaintext passwords, anyone?)

But "THEY made stupid decisions" is not a valid reason for YOU to make stupid decisions.

Atom is an excellent text editor. I love SO many things about it. Its integration with Windows is not one of the things I love, for the reasons discussed in this thread and for other reasons too (why does it have to be a major pain in the ass to associate a file extension with Atom? Oh, and upgrading silently breaks file associations too)

**Comment by stevesobol** [comment]
I would point out that if I had the time, I'd fork Atom and make better decisions regarding the installer. I don't have anywhere *near* enough time.

**Comment by lee-dohm** [comment]
Thanks @stevesobol for your feedback. If you want to be 100% in control of the installation and update process, you have that option. You can download the zip format for your platform and install it wherever you like. You can manage exactly when and how Atom updates in this way, Atom will not try to autoupdate when you've installed using the zip format.

We're also making progress on running Atom from a portable install, so you can install Atom to a USB drive and unplug it when you're not using it.

So while you may not approve of the choices that were made in designing the installer, there are options.

**Comment by Zireael07** [comment]
No autoupdater for the zip is a big nuisance, though. And the zip is not advertised enough.

**Comment by stevesobol** [comment]
@lee-dohm - Actually, I like that answer. It's a very good compromise between doing nothing, and building everything from source. I could create an installer using (for example) InnoSetup that installs the program into \Program Files (x86) and sets up file associations, etc. 

I was pretty harsh, wasn't I... let me just say that other than the silent updates, I agree that there are valid reasons for the design and packaging choices that have been made... I think I need to probably tone it down a little bit in the future... thank you for not responding to my snark with more snark. I honestly wasn't trying to be snarky (although I ended up succeeding, despite my best efforts).

@Zireael07 here's what I'm thinking. I do this, and include an Atom plugin that checks if a new version of my installer is available (I would update it when new versions of Atom are released), but I prompt the user to download and install, rather than updating silently.

@thx1200 @shawnr13 @rbcmhelpdesk  if I was to start packaging and making Atom releases available this way, in a way that won't cause problems with corporate group policies, would y'all be interested in using the fruits of my labor?

**Comment by stevesobol** [comment]
@lee-dohm Also, if I end up creating these alternate installers on a regular basis, would it be possible to get a link on atom.io somewhere? Probably not on the homepage; individuals and SMBs that don't use group policies won't need it, and putting the link there is likely to cause confusion. Some prominent place somewhere else on the site, though.

**Comment by lee-dohm** [comment]
> Also, if I end up creating these alternate installers on a regular basis, would it be possible to get a link on atom.io somewhere?

I'd be willing to add a link to them (making it very clear that they are unofficial) from [Discuss, the official Atom and Electron message board](https://discuss.atom.io). That should get sufficient :eyes: on them.

> I was pretty harsh, wasn't I

You were. I'm glad we worked things out though :grinning:

**Comment by stevesobol** [comment]
> I'd be willing to add a link to them (making it very clear that they are unofficial)

That'd be fine... just to get the links into Google. I will also make clear that I am solely responsible for supporting my work, and that it is not officially blessed by Github or by any of the Atom developers.

This is on my radar. I don't know when it will happen but I'm hoping I can make it happen soon.

Thanks :) and since I see upthread that I didn't apologize, I would like to apologize. I must have been having a pretty bad day - but that's no excuse. Won't happen again.

**Comment by lee-dohm** [comment]
> Thanks :) and since I see upthread that I didn't apologize, I would like to apologize. I must have been having a pretty bad day - but that's no excuse. Won't happen again.

It's quite alright. We all have bad days, believe me! Hopefully we just don't all have the *same* bad day :laughing:

> This is on my radar. I don't know when it will happen but I'm hoping I can make it happen soon.

Thanks for keeping us updated. Just let me know when you've got something :grinning:

**Comment by pmall** [comment]
> This also prevents users who aren't Administrator from installing the app at all.

That's why there is administrator : installing softwares for non administrator people.

User directory based install sucks a lot. Just put it in program files like any other software so it can be managable by windows server admins.

**Comment by stevesobol** [comment]
That's my goal with the alternate packaging. I haven't had any time to start working on it though. It's something I very badly want to do - but I'm not getting paid to do it - so it sits at the bottom of my list of priorities. :(

**Comment by Jtracy-ItPro** [comment]
I'm having a problem with this right now. I need to install this in several computer labs, and it must be there for all users. That's easy for the typical Windows app, you just install for the system and manage the updates. But the instructor is asking for Atom by name, unfortunately.

Notepad++ is workable in this scenario, so he'll probably have to settle for that or go someplace else.

It would be nice to keep the application files in "program files", and all user data in the profile folder. Admins could manage patches the way they manage other apps. I realize this flies in the face of the way modern dev tools work,. But these are public access PCs and will be managed as such.

Is there a supported way to install the app in the default user profile? Then it could be pre-installed for each user when their profile is built.

**Comment by Jtracy-ItPro** [comment]
I found a solution in packaging this app admin-style for all users.

Instead of the EXE, get the ZIP file. You have to look for it, but it's on the download page.
Don't even bother with the MSI, I know it's tempting.
Write a script to unpack the ZIP to %programdata%
All users have write access to this folder, and they can update freely. This would only get awkward if multiple users are logged in at the same time (terminal server / rds / etc)
In the same script, make shortcuts for Atom.EXE in start menu and/or desktop for all users.
To disable the auto-updater and/or enable the welcome screen, save a config file with your desired settings. Then put more code in the install script to sweep all profiles and copy the config to c:\users<user>\.atom\

This will be used in various public-access computer labs, most of them mobile (thus the auto-update disabler).

**Comment by idontgitit** [comment]
yeah, +1 on the support for system based installer (program files) as well.  so that it can be deployed in a corporate environment.  we're trying to burn into our base image for our fleet.  user profile based installs don't let us do that.  we had similar issues with other apps like dropbox, but they've since updated it when we've complained.  so i say....offer both like chrome and bluejeans :)

**Comment by lovebdsobuj** [comment]
Giving people an option where to install, will be good. I think. 
Automatic install on UserProfile ugly things.

**Comment by stevesobol** [comment]
@lovebdsobuj Yeah, it's not a big deal (in general!) for standalone computers, or even maybe computers running on a small AD network, but as you get into the bigger enterprises I can see where it'd be a problem.

Unfortunately, I haven't been able to start my "enterprise installer" project yet. It'll continue to be low-priority unless, of course, someone is willing to step up with some money to pay me to get it done. That said, I haven't forgotten about it.

**Comment by frosztbyte** [comment]
I apologize for my abruptness, but I'm holding back my frustration as much as possible. Not having *any* installation options is **SO** stupid, especially for a nice, well-designed piece of software like Atom. I mean, no offense, but you guys (the developers) present the editor like it's the *best* thing since ice cream was invented, allowing for extensibility with thousands of addon packages and a vibrant community. And I have to say, even though it's got one of these new-age "Web app/mobile app" UIs that I can't stand, and is written (at least partially) in my least favorite language, JavaScript, it's a damn good product, and this is why I switched from VS Code.

So why the **hell** would you spend *that* much time, effort, money and energy designing, coding, debugging, documenting and otherwise creating an awesome piece of software and then spend _ten_ little seconds writing a ludicrous "setup" program that believes every single computer, and thus every end-user, to be the same, as if we all have the same likes and preferences? Are we merely _numbers_ and not people?! Doing this is similar to spending three months working on a masterpiece oil painting and then shoving it half-assedly into a 35 cent beat up old picture frame! It makes no sense whatsoever. That's more like the kind of dirty tactics Micro$oft and crApple are so fond of, rather than a forward-thinking FOSS advocate team; forcing users to automatically install every update there is and taking away a user's choice to set up their box the way _they_ like. I mean, if I didn't know this was an open source project, I'd think it was created by Microsoft. You _do_ understand geeks like to tweak the heck out of their boxes and have orgasms over the tedium of reading hundreds of pages of documentation or having countless ways to customize our software, right? I assumed you were developers ...

I don't mean to sound ungrateful, and definitely intend **no** offense. Atom is a splendid tool, and I've no doubt it will continue to remain used by many. But if you don't stop treating your user base, which is composed primarily of programmers, power users and system admins, like a bunch of Kindergarteners, insulting our intelligence, you **WILL** eventually end up hemorrhaging losers, and your reputation will suffer as well.

This is a fact I've learned based on twenty-plus years of development experience, a master's in computer science and a lifetime of entrepreneurship and using computers. I'm not bragging. I'm simply giving some advice. But it's your project. If you don't like having a lot of users, or just like to read a lot of negative comments, by all means, please continue your asinine practice. Maybe I'll switch back to VS Code or Notepad++ when I use Windoze, or even just use Vim/Emacs like I do on Linux.

Anyway, thanks for Atom, anyway. There _is_ a lot of good about it, but much more room for improvement, and I don't mean the application itself.

**Comment by lee-dohm** [comment]
@frosztbyte Please see [my comment above](https://github.com/atom/atom/issues/5259#issuecomment-175237027). We do offer options by using the Zip file method. We know that people like to set up their computers many, many different ways. In so many different ways, as a matter of fact, that trying to include options for all of that in the installer would quickly get out of hand. So this system keeps the installer simple and much more maintenance-free so that we can spend more time working on the editor itself.
